samples: boards: Add readme for gpio counter.

jenmwms opened the pull request:
This adds a README.rst for the UP Squared board GPIO counter sample
using details in the application source code and giving context.
This offers overview, requirements, and instructions to build and
flash, in complement to the cited reference.

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

dbkinder commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,70 @@
+
+Build the sample in the following way:
+.. zephyr-app-commands::
+    :zephyr-app: samples/boards/up_squared/gpio_counter
```
Need a blank line before the directive (line 49) and previous sentence, and these parameters should be indented to the ``z`` in ``zephyr-app-commands``

jenmwms replied:
Thanks for clarifying!

dbkinder reviewed (changes requested):
one more tweak.  (Thanks for adding the doc!)

nashif commented:
> author email (jenmwms [jennifer.m.williams@intel.com](mailto:jennifer.m.williams@intel.com)) needs to match one of the signed-off-by entries.

You need to set your git firsname, lastname correctly

Right now you have 

```
commit d092082bd6a66e1013712e7bfba33af58cec730b (HEAD, upstream/pull/21190)
Author: jenmwms <jennifer.m.williams@intel.com>
Date:   Wed Dec 4 10:59:29 2019 -0800
```

jenmwms committed:
samples: boards: Add readme for gpio counter.

This adds a README.rst for the UP Squared board GPIO counter sample
using details in the application source code and giving context.
This offers overview, requirements, and instructions to build and
flash, in complement to the cited reference.

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

jenmwms commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,70 @@
+References
+**********
+
+- https://docs.zephyrproject.org/latest/boards/x86/up_squared/doc/index.html
```
Thanks for clarifying!

jenmwms commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,70 @@
+    :board: up_squared
+    :goals: build
+
+Prepare the boot device (USB storage drive) as described for the UP Squared.
```
Correct - west flash is not part of the workflow for this board (up2) at this time.

jenmwms replied:
I will add the reference (:ref: up_squared). Thanks!

jenmwms commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,70 @@
+
+Replace :code:`<tty_device>` with the port where the UP Squared board
+can be found. For example, under Linux, :code:`/dev/ttyUSB0`.
+The ``-b`` option sets baud rate.
```
Great suggestion. The 'button' to see the sample in action is a GPIO pin as input. I can add a brief description for implementation options (off-board analog button press for input, LEDs for visual output at the pins.

jenmwms commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,70 @@
+
++------------+-----+-----+-----+-----+-----+
+| Bit (bin)  | n/a |   3 |   2 |   1 |   0 |
++------------+-----+-----+-----+-----+-----+
```
Good point. It is not a header...but maybe I should add one to clarify. This shows the mapping between binary output of the value and the various GPIO pin numbers a developer should be aware of.

jenmwms commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,70 @@
+configurations (above) and then select to save the BIOS settings and reset.
+
+The board will reboot and then enter GRUB boot loader. Select to boot Zephyr and
+the board will start to execute the sample. Apply input to trigger the increment
```
Good point. I added a description now (based on other comment to include how this can be implemented to observe the sample behavior).

dbkinder commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,95 @@
+provided by a GPIO output (active High) pin of another GPIO device (eg Arduino
+Uno).The Up Squared GPIO output signals can each be connected to a simple LED
+circuit on a breakout breadboard to illuminate the 4-bit counter value, as
+shown in the example below:
```
The easiest fix to get rid of the doc build errors is to change the end of this line to be ``below::``  (adding a second colon says to treat the next indented section as a code block)

dbkinder commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,95 @@
+References
+**********
+
+- :ref: 'up_squared' board documentation
```
change to:  ``:ref:`up_squared` ``  (need to use back-ticks not apostrophes, and no extra spaces)

dbkinder commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,95 @@
+   :goals: build
+
+Prepare the boot device (USB storage drive) as described for the UP Squared
+board :ref: 'up_squared'. Insert the USB boot device containing the prepared
```
same fix here: ``:ref:`up_squared` ``

dbkinder commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,95 @@
+console display of the counter value. For example, the input signal can be
+implemented with an analog button on a breakout breadboard, or with a basic pulse
+provided by a GPIO output (active High) pin of another GPIO device (eg Arduino
+Uno).The Up Squared GPIO output signals can each be connected to a simple LED
```
add a space after the period (end of the sentence)

dbkinder commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,95 @@
+The sample enables a pin as GPIO input (active high) that triggers the increment
+of a counter (range is 0 to size of counter_pins[]). The counter increments for
+each change from 0 to 1 on HAT Pin 16 (BIOS Pin 19). The value of the counter is represented
+on GPIO output (active high) as a 4-bit value
```
While I think the intent was for the pins to be a 4-bit representation of the counter value, i.e., the pins go 0000, 0001, 0010, 0011, 0100, ..., 1111, it looks like the code masks the counter with 0b11 (for 4 pins) so there's an error in the code. Or I could be wrong...

jenmwms replied:
Thanks for asking.  I do not think there is an issue with that aspect (there are other known issues).

I tested the existing sample just now (cherry-picked 005aff7), and it behaves as 'intended' - the output pins display the count through a value of 0x0 to 0xf displayed in binary across the pins.  For example, 0xc (serial console) is represented by 0b1100. Scaling the sample to less or more output pins would be neat exercise...so, I need to update the range described here.

dbkinder reviewed (changes requested):
almost...

dbkinder reviewed (changes requested):
Getting closer...

dbkinder commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,97 @@
+.. _up_squared gpio_counter:
+
+UP Squared
```
Actually the title here should be ``UP Squared GPIO counter`` (and adjust the underline to be the same length as the new title).

jenmwms commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,97 @@
+Building and Running
+********************
+
+Build the sample in the following way::
```
Interesting. Thanks for helping point out the differences.

jenmwms commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,97 @@
+software binary of the sample.
+
+Connect the board to a host computer and open a serial connection for serial
+console interface::
```
I see, thank you!

jenmwms commented on samples/boards/up_squared/gpio_counter/README.rst:
```diff
@@ -0,0 +1,97 @@
+     :goals: build
+
+Prepare the boot device (USB storage drive) as described for the UP Squared
+board :ref: `up_squared`. Insert the USB boot device containing the prepared
```
Cool! Is the double-back-tick  an alternative to the suggestion in the [guide](https://docs.zephyrproject.org/latest/guides/documentation/index.html)?
Refer to the `show this text instead <internal-linking>`_ page

jenmwms replied:
I get syntax issues with the double-back-tick. The guide also suggests

:ref:`some other text <my label name>`

dbkinder replied:
oops, my typo:  ``... as described for the :ref:`up_squared` board.``

dbkinder reviewed (approved):
LGTM, thanks!
